compiler/optimizer/Inliner.cpp

```diff
@@ -5114,22 +5114,36 @@ bool TR_InlinerBase::inlineCallTarget2(TR_CallStack * callStack, TR_CallTarget *
-       && (tif->resultNode() == NULL || tif->resultNode()->getReferenceCount() == 0))
-      {
-      /**
-       * In OSR, we need to split block even for cases without virtual guard. This is
-       * because in OSR a block with OSR point must have an exception edge to the osrCatchBlock
-       * of correct callerIndex. Split the block here so that the OSR points from callee
-       * and from caller are separated.
-       *
-       * This will not uncommon the return value if the block is split, instead it expects a temporary
-       * to have been generated so that nothing is commoned. This is valid under the current
-       * inliner behaviour, as the existance of a catch block will result in the required temporary being created.
-       */
-      TR::Block * blockOfCaller = previousBBStartInCaller->getNode()->getBlock();
-      TR::Block * blockOfCallerInCalleeCFG = nextBBEndInCaller->getNode()->getBlock()->split(callNodeTreeTop, callerCFG);
-      callerCFG->copyExceptionSuccessors(blockOfCaller, blockOfCallerInCalleeCFG);
+   else if (comp()->getOption(TR_EnableOSR) && !comp()->osrInfrastructureRemoved())
+      {
+      TR::Block * blockOfNextBBEndInCaller = nextBBEndInCaller->getNode()->getBlock();
+      TR_OSRMethodData *osrMethodData = comp()->getOSRCompilationData()->findOSRMethodData(comp()->getCurrentInlinedSiteIndex(), calleeSymbol);
+
+      if (osrMethodData && osrMethodData->getOSRCatchBlock())
+         {
+         /**
+          * In OSR, we need to split block even for cases without virtual guard. This is
+          * because in OSR a block with OSR point must have an exception edge to the osrCatchBlock
+          * of correct callerIndex. Split the block here so that the OSR points from callee
+          * and from caller are separated.
+          *
+          * This will not uncommon the return value if the block is split, instead it expects a temporary
+          * to have been generated so that nothing is commoned. This is valid under the current
+          * inliner behaviour, as the existance of a catch block will result in the required temporary being created.
+          */
+
+         debugTrace(tracer(), "Splitting block to have distint exception edges for callee %d\n", comp()->getCurrentInlinedSiteIndex());
+
+         TR::Block * blockOfCallerInCalleeCFG = blockOfNextBBEndInCaller->split(callNodeTreeTop, callerCFG);
+         callerCFG->copyExceptionSuccessors(blockContainingTheCall, blockOfCallerInCalleeCFG);
+         }
+      else
+         {
+         debugTrace(tracer(), "Copy OSR exception edge from caller block_%d to block_%d\n", blockContainingTheCall->getNumber(), blockOfNextBBEndInCaller->getNumber());
+
+         // Need to copy the exception edge from caller block containing the call to the last callee block after merge
+         callerCFG->copyExceptionSuccessors(blockContainingTheCall, blockOfNextBBEndInCaller);
```

Comment:
The copy should not be bind - there should be an assert here that there isn't already an exception successor on the block where this is being copied to.

Reply:
Do you mean there isn't already an exception successor to OSR catch block? We copy exception edges except OSR exception edge from `blockContainingTheCall` to all the callee blocks that are not OSR catch block or OSR code block [here](https://github.com/eclipse/omr/blob/master/compiler/optimizer/Inliner.cpp#L5070). So there can be exception successor on the merged block at this point.

Reply:
So we need to make sure that the exception edges are consistent and at the very least there needs to be a fatal assert, but probably some conditional code that will abort the merge if both blocks have OSR exception successors or conflicting exception edges.

Reply:
`copyExceptionSuccessors` will not copy existing successors. IMO, I don't think we need any assert.
The merge happens [here](https://github.com/eclipse/omr/blob/master/compiler/optimizer/Inliner.cpp#L5022). The trees from the last callee block and the caller trees immediately after the call are merged, they are all put into `lastCalleeBlock`. Then all non-OSR exception successors from `blockContainingTheCall` are copied to every callee block. So `lastCalleeBlock` has all exception successors from the callee and non-OSR exception successors from the caller. Then in the code I touched, we call `copyExceptionSuccessors` to copy the rest of exception successors, which is the OSR exception successor. At this point, I don't think there are conflicting exception edges.

I can pass in a filter function to `copyExceptionSuccessors`  to only copy OSR exception successor to make the code clear.

Reply:
@andrewcraik Added code to assert if there is a mismatch between the osr meta data and CFG.